app/models/nic/base.rb

```diff
@@ -277,13 +280,6 @@ def to_audit_label
 
     protected
 
-    def normalize_mac
-      self.mac = Net::Validations.normalize_mac(mac)
-      true
-    rescue Net::Validations::Error => e
-      errors.add(:mac, e.message)
-    end
-
```

Comment:
What exactly wrong is with before validation? We use Rails callbacks a lot through the codebase, teach me.

Reply:
Rails model hooks cycle is as follow:

1. instancation (`new MyModel param: value, param2: value2, ...`)
2. _before_validation_: model attributes may be wrong and can be normalized here
3. validation
4. _before_save_ (maybe "after validation too"): model passed validation and is ready to be saved
5. save
6. _after_save_: save was successfull.

As we can see, using model data while validation did not happen is a risk to use invalid values in another part of the application (simple example of this:

```rb
post = Post.new title: '       some badly formatted garbage'
post.send_notification_emails
post.save # will trigger "before_validation" hooks which in this example cleans the title, but mails are sent using the wrong one.
```

The current Foreman's behavior about mac addresses is to normalize it on a _before_validation_ hook, which is fine, as long as the data is not used somewhere else, but as we encountered issues with mac addresses being upcased in the steps of a virtual machine creation, our guess was an issue with the mac being used before its validation (and normalization).

The minimal fix to this is to downcase macs in the database for consistency and downcase macs when they are assigned to the object (the `mac=` method).

The best approach would be to manage to perform validation earlier in the process (fail fast!) but this will need a good understanding of the mechanisms involved behind host creation, in order to reorder the sequence. It would be something like:

```rb
host Host.new(params)
host.validate
if host.valid?
  continue_machine_creation_process
else
  # raise or return the errors to the user
end
```

If things are done in the model, it's a better idea to do it in the `before_save` hook, as data should be valid at this point.

But as I say, this rework would need a good understanding and some time...

I hope these explanations helped!

Reply:
@mtancoigne Many thanks for detailed explanations.
@lzap @ekohl After this explanation, what do you suggest?
IMHO, 
 * API break is a no-go,
 * reorganize whole validation/normalization process requires a mentor and probably a sponsorship as its time consuming if we want to refactor this safely
 * the left way is to use the first tiny/poor patch I provide at first with the db migration added

Reply:
I agree with that, if we don't want to rework too much code, the small patch seems to be the solution (overriding `mac=` to downcase the value).

Reply:
It may be ok to override `mac=` but `normalize_mac` can raise an error if it's invalid. That's probably the main reason that the API change is there. Perhaps there needs to be a best effort normalize in the assignment while still allowing invalid data. Then a validation that does proper error handling.

And yes, the whole host orchestration is one area that deserves a lot of love. it can be very complex because a compute resource may provide a MAC address. That may be what you're running into.

I think that's happening here:
https://github.com/theforeman/foreman/blob/ceb276fbe96b97770f5d292b1eadca0205a34a0a/app/models/concerns/orchestration/compute.rb#L347-L349

That may also be a place to normalize, but I think it's wrong because it may lead to a mismatch. It's probably better for a compute provider to solve it so it always returns a good value. For example here: https://github.com/theforeman/foreman_fog_proxmox/blob/ca9475c15731e866cc3d8b1998320b30838f16d1/app/models/concerns/fog_extensions/proxmox/server.rb#L47-L49

What do you think?

Reply:
My guts are telling me this is not right, but I am not that skilled RoR developer so I am gonna pull @ares or @ezr-ondrej to assist here. What I particularly don't like is that this is not a problem of MAC field, but generic flaw of Ruby on Rails design. Also where do get the confidence that somewhere deep in Rails code it will use attribute setter to set the MAC instead of accessing an instance variable from?

I feel like fixing the MAC address in both UI and API controller is way cleaner - these are the main (perhaps all) input points where a human would enter a MAC address. But I am not the one to decide here. I do appreciate the effort and I do understand the importance of this bug to be fixed, thank you for patience with us.

Reply:
I believe this is flaw of our orchestration model, that is doing FAAAR too much in `around_save` and most of it `before_save` hook, where as @ekohl pointed out we allow changing the model (that's antipatern because it is after the validation have passed)
Two issues I see with this:
1. We can fail with exception on setting invalid value, that is quite bad
2. We do not validate the mac address anymore.

I agree with @ekohl probably the easiest would be normalizing in the setter the best effort (probably just the `.downcase` would be enough). Plus keep the validation on mac address.